### PR TITLE
BotData Shared Context

### PR DESCRIPTION
**Pull request by tausifmuzaffar**
We want to be able to sync data between bots when we specified its allowed to do it in the room settings.

The workflow is like this: Room Defines Bot1 must sync with Bot2 -> Bot1 BotData is created/updated -> we determine if we need to sync -> create/update Bot2 BotData

### CODE REVIEW

File: db/helpers/botDataUtils.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+const ONE = 1;
+const TWO = 2;
+
+/**
+ * Check if a string is a JSON
+ *
+ * @param {String} item - String to check
+ * @returns {Boolean} - if String is JSON
+ */
+function isJson(item) {
```
Comment (iamigo) [inline db/helpers/botDataUtils.js]: https://www.npmjs.com/package/is-json?

File: db/helpers/botDataUtils.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+  if (typeof itemParse === 'object' && itemParse !== null) {
+    return true;
+  }
+
+  return false;
+} // isJson
+
+/**
+ * Check if a string is a JSON
```
Comment (iamigo) [inline db/helpers/botDataUtils.js]: Wrong function comment for this function?

Comment (iamigo) [reply db/helpers/botDataUtils.js]: This function needs some explanation and examples.

File: db/helpers/botDataUtils.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      outputValue = outputValue
+        .replace(replaceString, instance.value);
+    }
+  }
+
+  return outputValue;
+} // replaceValue
+
+/**
+ * Determine if bot data needs to synced between bots and perform the action.
```
Comment (iamigo) [inline db/helpers/botDataUtils.js]: "needs to synced"??

File: db/helpers/botDataUtils.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+ *
+ * @param {Object} seq - The array to test
+ * @param {Object} instance - The array to test
```
Comment (iamigo) [inline db/helpers/botDataUtils.js]: is "instance" really the array to test, again?

File: db/helpers/botDataUtils.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+ * @returns {undefined} - OK
+ * @throws {validationError} - if the array does not contain valid attributes
+ */
+function updateValues(seq, instance) {
+  const Room = seq.models.Room;
+  const Bot = seq.models.Bot;
+  const BotData = seq.models.BotData;
+  const botNames = {};
+  const botsIds = {};
+  return Bot.findAll({
```
Comment (iamigo) [inline db/helpers/botDataUtils.js]: Bot.findAll with no where clause? I don't understand.

File: db/helpers/botDataUtils.js (hunk 1 above)
Comment (tausifmuzaffar) [reply db/helpers/botDataUtils.js]: I tried this package but it doesnt seem to be as good at the code i wrote it return false for this string

{ response: 'this is a {$TestBot1.response$}' }

Comment (iamigo) [reply db/helpers/botDataUtils.js]: Is `{$varName$}` a standard for doing variable replacement? Where is this used? Why not more standard `${varName}` or "template" style `{{varName}}`?

**Comment by iamigo** [comment]
jscs lint failures

File: db/helpers/botDataUtils.js (hunk 1 above)
Comment (tausifmuzaffar) [reply db/helpers/botDataUtils.js]: Some of the bots already use variable placements, like comms bot and pagerduty use handlebars. So we needed something unquie that would conflict with things

Comment (iamigo) [reply db/helpers/botDataUtils.js]: Could be confusing, hard for bot developers to grok non-standard syntax. What alternatives have you explored?

Comment (tausifmuzaffar) [reply db/helpers/botDataUtils.js]: Ive decided to switch ${} since handlebars use {{}}; and I dont know another JS template that use it withouth ``

**Comment by tausifmuzaffar** [comment]
@kamilsmuga @BrandonMurnane can yall review this?

**Pull request merged**